CRM/Contribute/Form/Contribution/Confirm.php

```diff
@@ -492,15 +495,18 @@ public function buildQuickForm() {
     // Make a copy of line items array to use for display only
     $tplLineItems = $this->_lineItem;
     if (CRM_Invoicing_Utils::isInvoicingEnabled()) {
-      // @todo $params seems like exactly the wrong place to get totalTaxAmount from
-      // this is a calculated variable so we it should be transparent how we
-      // calculated it rather than coming from 'params'
-      $this->assign('totalTaxAmount', $params['tax_amount']);
+      $taxAmount = 0;
+      foreach ($tplLineItems ?? [] as $lineItems) {
```

Comment:
I guess this is cos we have line items 'handy' - but I'd rather grab it from the db contribution than rely on whatever got through the form wrangling to this point

Reply:
I think there are a number of issues with "Separate membership payment" and this PR only scratches the surface.

If you look in confirm.tpl and thankyou.tpl you'll see there is special handling for `$is_separate_payment` which does not include displaying tax. But the tax is also not assigned properly to the form. In my view we should be getting rid of all that special handling and relying on the "lineitems" display in all cases (as we've already done with contributions in the backend).

@demeritcowboy The problems that you found, are they caused by this PR or pre-existing?

Reply:
Ok it looks pre-existing, so I see now this is just about replacing the $params part.

I don't have an opinion on where it should come from - maybe just if there's any hooks that would make a difference and one would handle that better.